File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.cpp
class Solution {
public:
    vector<int> restoreArray(vector<vector<int>>& ap) {
        
        unordered_map<int,vector<int>> mp;
        
        vector<int> ans;
        
        for(auto &i : ap) mp[i[0]].push_back(i[1]), mp[i[1]].push_back(i[0]);
        
        int st = 0;
        for(auto &i : mp){
            if(i.second.size()==1) {
                
                st = i.first;
                break;
            }
        } 
        
        int prev = 0;
        
        ans.push_back(st);
        prev = st;
        st = mp[st][0];
        
        while(mp[st].size() != 1){
            ans.push_back(st);
            int tot = mp[st][0] + mp[st][1];
            int next = tot-prev;
            prev = st;
            st = next;
        }
        
        ans.push_back(st);
        return ans;
    }
};

File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.py
class Solution:
    def restoreArray(self, ap: List[List[int]]) -> List[int]:
        
        ans = list()
        mp = dict()
        
        for i in ap:
            one, two = i[0], i[1]
            mp.setdefault(one,[]).append(two)
            mp.setdefault(two,[]).append(one) 
            
        
        st, prev = 0, 0
        # print(mp)
        for k,v in mp.items():
            if len(v) == 1:
                st = k 
                break 
        
        ans.append(st)
        prev = st 
        st = mp[st][0] 
        
        while len(mp[st]) != 1:
            ans.append(st)
            tot = sum(mp[st])
            nxt = tot-prev 
            prev = st 
            st = nxt 
        
        ans.append(st)
        return ans
            
        